text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+             | "e"      // str
+             | "f"      // f32
+             | "h"      // u8
+             | "i"      // isize
+             | "j"      // usize
+             | "l"      // i32
+             | "m"      // u32
+             | "n"      // i128
+             | "o"      // u128
+             | "s"      // i16
+             | "t"      // u16
+             | "u"      // ()
+             | "v"      // ...
+             | "x"      // i64
+             | "y"      // u64
+             | "z"      // !
+
+// If the "U" is present then the function is `unsafe`.
+// If the "J" is present then it is followed by the return type of the function.
+<function-type> := "F" ["U"] [<abi>] {<type>} ["J" <type>] "E"
+
+<abi> = "K" (
+           "d" |   // Cdecl
+           "s" |   // Stdcall
+           "f" |   // Fastcall
+           "v" |   // Vectorcall
+           "t" |   // Thiscall
+           "a" |   // Aapcs
+           "w" |   // Win64
+           "x" |   // SysV64
+           "k" |   // PtxKernel
+           "m" |   // Msp430Interrupt
+           "i" |   // X86Interrupt
+           "g" |   // AmdGpuKernel
+           "c" |   // C
+           "x" |   // System
+           "r" |   // RustCall
+           "j" |   // RustInstrinsic
+           "p" |   // PlatformInstrinsic
+           "u"     // Unadjusted
+        )
+
+<disambiguator> = "s" [<base-62-digit>] "_"
+
+<generic-arguments> = "I" {<type>} "E"
+
+<substitution> = "S" [<base-62-digit>] "_"
+
+// We use <path-prefix> here, so that we don't have to add a special rule for
+// compression. In practice, only <identifier> is expected.
+<instantiating-crate> := <path-prefix>
+```
+
+### Punycode Identifiers
+
+Punycode generates strings of the form `([[:ascii:]]+-)?[[:alnum:]]+`. This is problematic for two reasons:
+
+- Generated strings can contain a `-` character; which is not in the supported character set.
+- Generated strings can start with a digit; which makes them clash with the byte-count prefix of the `<identifier>` production.
+
+For these reasons, vanilla Punycode string are further encoded during mangling:
+
+- The `-` character is simply replaced by a `_` character.
+- The part of the Punycode string that encodes the non-ASCII characters is a base-36 number, using `[a-z0-9]` as its "digits". We want to get rid of the decimal digits in there, so we simply remap `0-9` to `A-J`.
+
+Here are some examples:
+
+| Original        | Punycode        | Punycode + Encoding |
+|-----------------|-----------------|---------------------|
+| føø             | f-5gaa          | f_Fgaa              |
+| α_ω             | _-ylb7e         | __ylbHe             |
+| 铁锈             | n84amf          | nIEamf              |
+| 🤦              | fq9h            | fqJh                |
+| ρυστ            | 2xaedc          | Cxaedc              |
+
+With this post-processing in place the Punycode strings can be treated like regular identifiers and need no further special handling.
+
+
+## Compression
```

Comment:
I'd be very tempted to omit this kind of ad-hoc compression scheme in favour of using a standard algorithm like `zstd` (say, zstd with a well-defined domain-specific dictionary). (cc @Cyan4973 - are there any examples of using zstd for per-symbol compression?)

Historically C++ demanglers have been very fragile, and I suspect a big part of that is due to the implementation complexity of the Itanium ABI compression mechanism.

Aside from the implementation issues, because this is so coupled with the definition of the mangling scheme itself, it means that any future evolution of mangling needs to also take compression into account. Using a completely separate compression layer makes this a non-issue. The other nice thing about making compression largely isolated from the rest of the encoding is that it means it can be added in a second pass as an extension once we have some experience with uncompressed mangling - maybe it wouldn't be so bad?

The main problem with using an external library is that any Rust demangler introduces another dependency. This is particularly worth considering when integrating Rust demangler support into other tools like binutils/llvm/perf/valgrind/etc.

Reply:
I believe we can come up with a simple compression scheme (i.e. refer back to the byte position where the first occurrence of something was encoded).

This would allow a demangler implementation to have 0 external dependencies, and the specification would also not implicitly depend on another standard.

It might also behave better than `zstd` given the limitation of `[a-zA-Z0-9_]` (which makes bit streams less appealing), and it has the advantage that any path component name is guaranteed to show up *in clear*.

*However*, I would not be opposed to at least having a compiler mangling option which disregards the `[a-zA-Z0-9_]` limitation and which does the best compression it can, for use in situations where that might be advantageous (although at that point, you might be better off with symbol names just, say, hashes, and keep everything else in split debuginfo, and/or an ad-hoc mapping from hashes to symbol names).

Oh and should definitely gather data on all compression schemes we can think of (that are not too painful to implement), before we accept the RFC!

Reply:
I've listed the `zstd` option as point 5 in [Rational and Alternatives](https://github.com/michaelwoerister/rfcs/blob/symbol-mangling/text/0000-symbol-name-mangling-v2.md#rationale-and-alternatives). I would be interested in seeing how `zstd` performs. But it does come with some real downsides:
- Every demangler would have to support `zstd`. That's another dependency that not everyone might want to pull in. 
- The specification of the mangling scheme would depend on the specification of `zstd`. I see that there's an [IETF RFC](https://tools.ietf.org/html/rfc8478) for it. That's good. But it's still rather heavyweight. 
- Mangled symbol would not retain any human-readability at all.

I think one of the next steps would be to collect a body of symbol names for testing different compression schemes.

Reply:
@jsgf I do acknowledge, btw, that an AST-independent compression scheme is clearly beneficial when it comes to evolving the grammar.

Reply:
@michaelwoerister Yes, I think extra dependencies and non-readability reasonable counter-arguments to using zstd.

(waving hands wildly) I'm assuming that we wouldn't bother compressing small symbols, so they would remain directly readable, and large symbols with any compression scheme would be such a soup that even if the compression scheme leaves some parts "in the clear" they're still not directly readable in any practical sense.

Reply:
zstd and similar bit-compression algorithms are not very suitable for compressing short strings at all, and encoding the compressed bit-stream into `[a-zA-Z0-9]` (required by toolchains) would lose most of the compression again.